src/measure_ops.jl

```diff
@@ -103,18 +103,89 @@ function eigenbasis(op::YGate)
 end
 
 function measure!(
-    postprocess::PostProcess,
+    postprocess::NoPostProcess,  # operator measuring does not allow removing bits or resetting vaules.
     op::AbstractBlock,
     reg::AbstractRegister,
     locs::AllLocs;
     kwargs...,
 )
     _check_msize(op, reg, locs)
     E, V = eigenbasis(op)
-    res = measure!(postprocess, ComputationalBasis(), reg |> V', locs; kwargs...)
-    res2 = measure!(postprocess, ComputationalBasis(), reg, locs; kwargs...)
-    postprocess isa NoPostProcess && apply!(reg, V)
-    diag(mat(E))[Int64.(res).+1]
+    res = measure!(postprocess, BlockedBasis(diag(mat(E))), reg |> V', locs; kwargs...)
+    apply!(reg, V)
+    return res
+end
+
+struct BlockedBasis{VT}
+    perm::Vector{Int}
+    values::VT
+    block_ptr::Vector{Int}
+end
+
+subblock(bb::BlockedBasis, i::Int) = bb.block_ptr[i]:bb.block_ptr[i+1]-1
+nblocks(bb::BlockedBasis) = length(bb.block_ptr)-1
+
+function BlockedBasis(values::AbstractVector{T}) where T
+    if length(values) == 1
+        return BlockedBasis([1], values, [1,2])
+    elseif length(values) == 0
+        return BlockedBasis([], values, [1])
+    end
+    order = sortperm(values; by=real)
+    values = values[order]
+    vpre = values[1]
+    block_ptr = [1]
+    unique_values = [vpre]
+    k = 1
+    for i=2:length(values)
+        v = values[i]
+        if !isapprox(v, vpre)  # use approx in order to ignore the round off error
+            k+=1
+            push!(block_ptr, i)
+            push!(unique_values, v)
+        end
+        vpre = v
+    end
+    push!(block_ptr, length(values)+1)
+    return BlockedBasis(order, unique_values, block_ptr)
+end
+
+function YaoBase.measure!(
+    ::NoPostProcess,
+    bb::BlockedBasis,
+    reg::ArrayReg{B,T},
+    ::AllLocs;
+    rng::AbstractRNG = Random.GLOBAL_RNG,
+) where {B,T}
+    state = (reg |> rank3)[bb.perm,:,:]  # permute to make eigen values sorted
+    pl = dropdims(sum(abs2, state, dims = 2), dims = 2)
+    pl_block = zeros(eltype(pl),nblocks(bb),B)
+    for ib=1:B
+        for i=1:nblocks(bb)
+            for k in subblock(bb, i)
+                pl_block[i,ib] += pl[k,ib]
+            end
+        end
+    end
+    res = Vector{Int}(undef, B)
+    @inbounds for ib = 1:B
+        ires = sample(rng, 1:nblocks(bb), Weights(view(pl_block,:,ib)))
+        # notice ires is `BitStr` type, can be use as indices directly.
+        range = subblock(bb,ires)
+        view(state,range,:,ib) ./= sqrt(pl_block[ires, ib])
+        view(state,1:range.start-1,:,ib) .= zero(T)
+        view(state,range.stop+1:size(state,1),:,ib) .= zero(T)
+        res[ib] = ires
+    end
```

Comment:
I think this is more readable

```suggestion
    @inbounds @views for ib = 1:B
        ires = sample(rng, 1:nblocks(bb), Weights(view(pl_block,:,ib)))
        # notice ires is `BitStr` type, can be use as indices directly.
        range = subblock(bb,ires)
        state[range,:,ib] ./= sqrt(pl_block[ires, ib])
        state[1:range.start-1,:,ib] .= zero(T)
        state[range.stop+1:size(state,1),:,ib] .= zero(T)
        res[ib] = ires
    end
```